### PR TITLE
Make config.vm.box optional or use to specify the template

### PR DESCRIPTION
**Pull request by bheuvel**
If not explicitly specified via the plugins template_name, will use Vagrants config.vm.box
Related to feature request #79

### CODE REVIEW

File: lib/vagrant-cloudstack/action/run_instance.rb [hunk 1]
```diff
@@ -5,6 +5,7 @@
 require 'vagrant-cloudstack/model/cloudstack_resource'
 require 'vagrant-cloudstack/service/cloudstack_resource_service'
 
+
```
Comment (miguelaferreira) [inline lib/vagrant-cloudstack/action/run_instance.rb]: not needed... doesn't hurt either

File: lib/vagrant-cloudstack/action/run_instance.rb [hunk 2]
```diff
@@ -34,7 +35,7 @@ def call(env)
           @network          = CloudstackResource.new(domain_config.network_id, domain_config.network_name, 'network')
           @service_offering = CloudstackResource.new(domain_config.service_offering_id, domain_config.service_offering_name, 'service_offering')
           @disk_offering    = CloudstackResource.new(domain_config.disk_offering_id, domain_config.disk_offering_name, 'disk_offering')
-          @template         = CloudstackResource.new(domain_config.template_id, domain_config.template_name, 'template')
+          @template         = CloudstackResource.new(domain_config.template_id, domain_config.template_name || env[:machine].config.vm.box, 'template')
```
Comment (miguelaferreira) [inline lib/vagrant-cloudstack/action/run_instance.rb:37]: Is this all it takes? I mean, isn't vagrant going to be looking for a vagrant box with that name?

Comment (bheuvel) [inline lib/vagrant-cloudstack/action/run_instance.rb:37]: Apparently not as this works on my laptop :wink:
I always specify the box_url to the "dummy" box in the Vagrant file:
`config.vm.box_url = "https://github.com/schubergphilis/vagrant-cloudstack/raw/master/dummy.box"`
this way I don't have to separately download the box each time/sometimes/first time, and Vagrant probably directly uses the url to get/check the dummy box.

On the other hand if I have a succeful run and I remove the box (`vagrant box remove dummy`) and url reference, it still works :confused: 

But now you mention it, I think that specifying the box_url in the Vagrant file might be a requirement for this to (allways) work.

Comment (miguelaferreira) [inline lib/vagrant-cloudstack/action/run_instance.rb:37]: I guess we need to go deep into vagrant with this, and make it somehow disregard the `config.vm.box` attribute, or somehow (not sure how and even why) make it work with the cloudstack templates.

I think you probably already have a better idea about what does vagrant actually do with that parameter. From everything it does we need to cut some of it off (and delegate to cloudstack) and keep the rest, probably with adaptations.

(Just brainstorming here...) Does that make any sense?

Comment (bheuvel) [inline lib/vagrant-cloudstack/action/run_instance.rb:37]: Just created a new VM, installed Vagrant plugin and all stuff.
Created a Vagrantfile without the box_url, "up"-ed the VM without problems and without requiring or downloading a "dummy" box...

Was triggered by https://github.com/mitchellh/vagrant-rackspace/blame/master/README.md#L52 .

Comment (miguelaferreira) [inline lib/vagrant-cloudstack/action/run_instance.rb:37]: Cool! We should write a functional test for this.
Next sprint (two weeks from now) I will take a day to work on that

File: lib/vagrant-cloudstack/plugin.rb [hunk 3]
```diff
@@ -24,7 +24,7 @@ class Plugin < Vagrant.plugin("2")
         Config
       end
 
-      provider(:cloudstack, parallel: true) do # Setup logging and i18n
+      provider(:cloudstack, { parallel: true, box_optional:  true}) do # Setup logging and i18n
```
Comment (bheuvel) [inline lib/vagrant-cloudstack/plugin.rb:27]: This setting at least removes the need to specify the dummy box in the first place.

**Comment by miguelaferreira** [comment]
LGTM

``` bash
cd functional-tests
./run_tests.sh 
::::>> Testing rsync
  ::::>> Testing with Vagrantfile.advanced_networking
 ✓ current directory is rsynced to VM

1 test, 0 failures
::::>> Testing vmlifecycle
  ::::>> Testing with Vagrantfile.advanced_networking
 ✓ create and destroy vm

1 test, 0 failures
```

**Pull request merged**